runtime/parachains/src/session_info.rs

```diff
@@ -0,0 +1,253 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Polkadot.
+
+// Polkadot is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Polkadot is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
+
+//! The session info module provides information about validator sets
+//! from prior sessions needed for approvals and disputes.
+//!
+//! See https://w3f.github.io/parachain-implementers-guide/runtime/session_info.html.
+
+use primitives::v1::{AuthorityDiscoveryId, SessionIndex, SessionInfo};
+use frame_support::{
+	decl_storage, decl_module, decl_error,
+	weights::Weight,
+};
+use crate::{configuration, paras, scheduler};
+
+pub trait Trait:
+	frame_system::Trait
+	+ configuration::Trait
+	+ paras::Trait
+	+ scheduler::Trait
+	+ AuthorityDiscoveryTrait
+{
+}
+
+decl_storage! {
+	trait Store for Module<T: Trait> as ParaSessionInfo {
+		/// The earliest session for which previous session info is stored.
+		EarliestStoredSession get(fn earliest_stored_session): SessionIndex;
+		/// Session information in a rolling window.
+		/// Should have an entry in range `EarliestStoredSession..=CurrentSessionIndex`.
+		Sessions get(fn session_info): map hasher(identity) SessionIndex => Option<SessionInfo>;
```

Comment:
A couple of questions:
1. Is identity hasher here ok? I assume the values will be consecutive ints.
2. Do we need init it on genesis? With `add_extra_genesis`?

Reply:
Re 1: Should be fine. After all, the window span is going to be on order of tens at the very most.

Re 2: I don't think this matters a lot. We just have to make sure that the client code doesn't fail spectacularly should they need the zeroth session. The running chains are long past the genesis, but they may bump into a similar issue with non genesis block anyway. However, since that's the public API, we should really reflect this in the docs.